File: Language/C#/Book/tmp/LINQ实战/Chapter05.VB/Chapter05/FormConditionalQuery.vb
Imports LinqInAction.LinqBooks.Common

Public Class FormConditionalQuery

  Public Sub New()

    ' This call is required by the Windows Form Designer.
    InitializeComponent()

    ' Add any initialization after the InitializeComponent() call.
    cbxPageCount.DisplayMember = "Key"
    cbxPageCount.ValueMember = "Value"
    cbxPageCount.DataSource = New KeyValuePair(Of String, Integer?)() { _
      New KeyValuePair(Of String, Integer?)("any", Nothing), _
      New KeyValuePair(Of String, Integer?)("at least 100", 100), _
      New KeyValuePair(Of String, Integer?)("at least 200", 200), _
      New KeyValuePair(Of String, Integer?)("at least 300", 300) _
    }

        cbxPageCount.SelectedIndex = 0
    cbxSortOrder.SelectedIndex = 0

    dataGridView.AutoGenerateColumns = False
    dataGridView.Columns.Clear()
    dataGridView.Columns.Add(New DataGridViewTextBoxColumn() With {.DataPropertyName = "Title", .HeaderText = "Title"})
    dataGridView.Columns.Add(New DataGridViewTextBoxColumn() With {.DataPropertyName = "PageCount", .HeaderText = "Pages"})
    dataGridView.Columns.Add(New DataGridViewTextBoxColumn() With {.DataPropertyName = "Publisher", .HeaderText = "Publisher"})
  End Sub

  Private Sub ConditionalQuery(Of TSortKey)(ByVal minPageCount As Integer?, ByVal titleFilter As String, ByVal sortSelector As Func(Of Book, TSortKey))
    Dim query As IEnumerable(Of Book)

    query = SampleData.Books
    If minPageCount.HasValue Then
      query = query.Where(Function(book) book.PageCount >= minPageCount.Value)
    End If
    If Not String.IsNullOrEmpty(titleFilter) Then
      query = query.Where(Function(book) book.Title.Contains(titleFilter))
    End If
    If Not sortSelector Is Nothing Then
      query = query.OrderBy(sortSelector)
    End If
    Dim completeQuery = query.Select(Function(book) New With {book.Title, book.PageCount, .Publisher = book.Publisher.Name})

    dataGridView.DataSource = completeQuery.ToList()
  End Sub

  Private Sub btnSearch_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnSearch.Click
    Dim minPageCount As Integer?
    Dim titleFilter As String

    minPageCount = CType(cbxPageCount.SelectedValue, Integer?)
    titleFilter = txtTitleFilter.Text
    If cbxSortOrder.SelectedIndex = 1 Then
      ConditionalQuery(minPageCount, titleFilter, Function(book) book.Title)
    ElseIf cbxSortOrder.SelectedIndex = 2 Then
      ConditionalQuery(minPageCount, titleFilter, Function(book) book.Publisher.Name)
    ElseIf cbxSortOrder.SelectedIndex = 3 Then
      ConditionalQuery(minPageCount, titleFilter, Function(book) book.PageCount)
    Else
      ConditionalQuery(Of Object)(minPageCount, titleFilter, Nothing)
    End If
  End Sub
End Class

File: Language/C#/Book/tmp/LINQ实战/Chapter05.CSharp/Chapter05/FormConditionalQuery.cs
#region Imports

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Linq.Expressions;
using System.Text;
using System.Windows.Forms;

using LinqInAction.LinqBooks.Common;

#endregion Imports

namespace LinqInAction.Chapter05
{
    public partial class FormConditionalQuery : Form
    {
        public FormConditionalQuery()
        {
            InitializeComponent();

            cbxPageCount.DisplayMember = "Key";
            cbxPageCount.ValueMember = "Value";
            cbxPageCount.DataSource = new KeyValuePair<String, int?>[] {
                new KeyValuePair<String, int?>("any", null),
                new KeyValuePair<String, int?>("at least 100", 100),
                new KeyValuePair<String, int?>("at least 200", 200),
                new KeyValuePair<String, int?>("at least 300", 300)
            };

            cbxPageCount.SelectedIndex = 0;
            cbxSortOrder.SelectedIndex = 0;

            dataGridView.AutoGenerateColumns = false;
            dataGridView.Columns.Clear();
            dataGridView.Columns.Add(new DataGridViewTextBoxColumn() { DataPropertyName = "Title", HeaderText = "Title" });
            dataGridView.Columns.Add(new DataGridViewTextBoxColumn() { DataPropertyName = "PageCount", HeaderText = "Pages" });
            dataGridView.Columns.Add(new DataGridViewTextBoxColumn() { DataPropertyName = "Publisher", HeaderText = "Publisher" });
        }

        private void ConditionalQuery<TSortKey>(int? minPageCount, String titleFilter, Func<Book, TSortKey> sortSelector)
        {
            IEnumerable<Book> query;

            query = SampleData.Books;
            if (minPageCount.HasValue)
                query = query.Where(book => book.PageCount >= minPageCount.Value);
            if (!String.IsNullOrEmpty(titleFilter))
                query = query.Where(book => book.Title.Contains(titleFilter));
            if (sortSelector != null)
                query = query.OrderBy(sortSelector);
            var completeQuery = query.Select(book => new { book.Title, book.PageCount, Publisher = book.Publisher.Name });

            dataGridView.DataSource = completeQuery.ToList();
        }

        private void btnSearch_Click(object sender, EventArgs e)
        {
            int? minPageCount;
            string titleFilter;

            minPageCount = (int?)cbxPageCount.SelectedValue;
            titleFilter = txtTitleFilter.Text;
            if (cbxSortOrder.SelectedIndex == 1)
                ConditionalQuery(minPageCount, titleFilter, book => book.Title);
            else if (cbxSortOrder.SelectedIndex == 2)
                ConditionalQuery(minPageCount, titleFilter, book => book.Publisher.Name);
            else if (cbxSortOrder.SelectedIndex == 3)
                ConditionalQuery(minPageCount, titleFilter, book => book.PageCount);
            else
                ConditionalQuery<Object>(minPageCount, titleFilter, null);
        }
    }
}